Docs additions, field common methods (#334)
yiisoft/form: docs/guide/en/field-methods.md

    <input type="text">
</div>
```

No ID is used by default.

### `containerClass()` / `addContainerClass()`

HTML class for outer container that wraps the field. In case of multiple classes, pass them as separate arguments.

```php
echo \Yiisoft\Form\Field\Text::widget()->containerClass('field-container', 'focus');
```

```html
<div class="field-container focus">
    <input type="text">
</div>
```

No class is used by default.

To add classes to existing ones, instead of replacing, use `addContainerClass()` method. In case of multiple classes, 
pass them as separate arguments.

```php
echo \Yiisoft\Form\Field\Text::widget()
    ->containerClass('field-container')
    ->addContainerClass('focus', 'primary');
```

Result:

```html
<div class="field-container focus primary">
    <input type="text">
</div>
```

### `useContainer()`

Whether to use outer container that wraps the field.

To disable container:

```php
echo Yiisoft\Form\Field\Text::widget()->useContainer(false);
```

```html
<input type="text">
```

Enable container (default):

```php
echo \Yiisoft\Form\Field\Text::widget()->useContainer();
```

## `PartsField` based fields

Parts field consists of 4 elements:

> - Label;
> - Hint;
> - Input;
> - Error.

Reviewer: ```suggestion
- label;
- hint;
- input;
- error.
```

Input here can be composite and can contain multiple elements (parts).

Built-in fields that extend from `PartsField`:

- [Button](fields/button.md)
- [ButtonGroup](fields/button-group.md)
- [Checkbox](fields/checkbox.md)
- [CheckboxList](fields/checkbox-list.md)
- [Date](fields/date.md)
- [DateTimeLocal](fields/date-time-local.md)
- [Email](fields/email.md)
- [Fieldset](fields/fieldset.md)
- [File](fields/file.md)
- [Hidden](fields/hidden.md)
- [Image](fields/image.md)
- [Number](fields/number.md)
- [Password](fields/password.md)
- [RadioList](fields/radio-list.md)
- [Range](fields/range.md)
- [ResetButton](fields/reset-button.md)
- [Select](fields/select.md)
- [SubmitButton](fields/submit-button.md)
- [Telephone](fields/telephone.md)
- [Text](fields/text.md)
- [Textarea](fields/textarea.md)
- [Time](fields/time.md)
- [Url](fields/url.md)

### `label()`

By default, label is not shown. To show it, call self-titled methods:

```php
echo \Yiisoft\Form\Field\Text::widget()->label('Name');
```

```html
<div>
    <label>Name</label>
    <input type="text">
</div>
```

> Another way is to use [config].

Reviewer: Is missed link?

### `hint()`

By default, hint is not shown. To show it, call self-titled methods:

```php
echo \Yiisoft\Form\Field\Text::widget()->hint('Enter name');
```

```html
<div>
    <input type="text">
    <div>Enter name</div>
</div>
```

> Another way is to use [config].

Reviewer: Is missed link?

### `error()`

By default, error is not shown. To show it, call self-titled methods:

```php
echo \Yiisoft\Form\Field\Text::widget()->error('Name is not valid.');
```

```html
<div>
    <input type="text">
    <div>Name is not valid.</div>
</div>
```

> Another way is to use [config].

Reviewer: Is missed link?

### `hideLabel()`

For the label, there is additional method to control its visibility. It can be hidden:

```php
use Yiisoft\Form\Field\Text;

$field = Text::widget();

/** @var bool $condition */
if ($condition) {
    $field = $field->hideLabel();
}

$field = $field->label('Name')

echo $field;
```

> The order of method calls is not important.

To show it:

```php
use Yiisoft\Form\Field\Text;

$field = Text::widget();

/** @var bool $condition */
if ($condition) {
    $field = $field->hideLabel(false);
}

$field = $field->label('Name')

echo $field;
```

### `labelConfig()`

> Config with [definitions](https://github.com/yiisoft/definitions) for label.

Reviewer: ```suggestion
Config with [definitions](https://github.com/yiisoft/definitions) for `Label` widget.
```

Usage:

```php
echo \Yiisoft\Form\Field\Text::widget()->labelConfig(['content()' => ['Name'], 'class()' => ['label']]);    
```

```html
<div>
    <label class="label">Name</label>
    <input type="text">
</div>
```

### `hintConfig()`

> Config with [definitions](https://github.com/yiisoft/definitions) for hint.

Reviewer: ```suggestion
Config with [definitions](https://github.com/yiisoft/definitions) for `Hint` widget.
```

Usage:

```php
echo \Yiisoft\Form\Field\Text::widget()->hintConfig(['content()' => ['Enter name'], 'class()' => ['hint']]);
```

```html
<div>
    <input type="text">
    <div class="hint">Enter name</div>
</div>
```

### `errorConfig()`

> Config with [definitions](https://github.com/yiisoft/definitions) for error.

Reviewer: ```suggestion
Config with [definitions](https://github.com/yiisoft/definitions) for `Error` widget.
```

Usage:

```php
echo \Yiisoft\Form\Field\Text::widget()->errorConfig(['content()' => ['Enter name'], 'class()' => ['hint']]);
```

```html
<div>
    <input type="text">
    <div class="error">Enter name</div>
</div>
```

### `labelAttributes()` / `addLabelAttributes()` 

HTML attributes for label.

```php
echo \Yiisoft\Form\Field\Text::widget()
    ->label('Name')
    ->labelAttributes(['class' => 'label', 'data-type' => 'name']);
```

> > The label must be visible (use [`label()`](#label) or [`labelConfig()](#labelconfig) methods).

Reviewer: ```suggestion
> The label must be visible (use [`label()`](#label) or [`labelConfig()`](#labelconfig) methods).
```

```html
<div>
    <label class="label" data-type="name">Name</label>
    <input type="text">
</div>
```

To add attributes to the existing ones instead of replacing, use `addLabelAttributes()` method. Note that values within
the same attribute will not be merged, newly added value overrides previous one.

```php
echo \Yiisoft\Form\Field\Text::widget()
    ->label('Name')
    ->labelAttributes(['class' => 'label', 'data-type' => 'name'])
    ->addLabelAttributes(['class' => 'focus', 'data-sort' => 1]);
```

Result:

```html
<div>
    <label class="focus" data-type="label" data-sort="1">Name</label>
    <input type="text">
</div>
```

### `hintAttributes()` / `addHintAttributes()`

HTML attributes for hint.

```php
echo \Yiisoft\Form\Field\Text::widget()
    ->hint('Enter name')
    ->hintAttributes(['class' => 'hint', 'data-type' => 'name']);
```

> > The hint must be visible (use [`hint()`](#hint) or [`hintConfig()](#hintconfig) methods).

Reviewer: ```suggestion
> The hint must be visible (use [`hint()`](#hint) or [`hintConfig()`](#hintconfig) methods).
```

</div>
```

### `errorAttributes()` / `addErrorAttributes()`

HTML attributes for error.

```php
echo \Yiisoft\Form\Field\Text::widget()
    ->error('Name is not valid.')
    ->errorAttributes(['class' => 'error', 'data-type' => 'name']);
```

> The error must be visible (use [`error()`](#error) or [`errorConfig()](#errorconfig) methods).

```html
<div>
    <input type="text">
    <div class="error" data-type="name">Name is not valid.</div>
</div>
```

To add attributes to the existing ones instead of replacing, use `addErrorAttributes()` method. Note that values within
the same attribute will not be merged, newly added value overrides previous one.

```php
echo \Yiisoft\Form\Field\Text::widget()
    ->errir('Enter name')
    ->errorAttributes(['class' => 'error', 'data-type' => 'name'])
    ->addErrorAttributes(['class' => 'focus', 'data-sort' => 1]);
```

Result:

```html
<div>
    <input type="text">
    <div class="error" data-type="name" data-sort="1">Enter name</div>
</div>
```

### `labelId()`

HTML ID for label.

```php
echo \Yiisoft\Form\Field\Text::widget()
    ->label('Name')
    ->labelId('label');
```

> > The label must be visible (use [`label()`](#label) or [`labelConfig()](#labelconfig) methods).

Reviewer: ```suggestion
> The label must be visible (use [`label()`](#label) or [`labelConfig()`](#labelconfig) methods).
```

```html
<div>
    <label id="label">Name</label>
    <input type="text">
</div>
```

### `hintId()`

HTML ID for hint.

```php
echo \Yiisoft\Form\Field\Text::widget()
    ->hint('Enter name')
    ->hintId('hint');
```

> > The hint must be visible (use [`hint()`](#hint) or [hintlConfig()](#hintconfig) methods).

Reviewer: ```suggestion
> The hint must be visible (use [`hint()`](#hint) or [`hintConfig()`](#hintconfig) methods).
```

Result:

```html
<div>
    <input type="text">
    <div id="hint">Enter name</div>
</div>
```

> 
> ```labelClass```

Reviewer: ```suggestion
```

### `errorId()`

HTML ID for error.

```php
echo \Yiisoft\Form\Field\Text::widget()
    ->error('Name is not valid.')
    ->errorId('error');
```

> > The error must be visible (use [`error()`](#error) or [errorlConfig()](#errorconfig) methods).

Reviewer: ```suggestion
> The error must be visible (use [`error()`](#error) or [`errorConfig()`](#errorconfig) methods).
```

Result:

```html
<div>
    <input type="text">
    <div id="error">Name is not valid.</div>
</div>
```

### `labelClass()` / `addLabelClass()`

HTML class for label. In case of multiple classes, pass them as separate arguments.

```php
echo \Yiisoft\Form\Field\Text::widget()
    ->label('Name')
    ->labelClass('label', 'focus');
```

> > The label must be visible (use [`label()`](#label) or [`labelConfig()](#labelconfig) methods).:

Reviewer: ```suggestion
> The label must be visible (use [`label()`](#label) or [`labelConfig()`](#labelconfig) methods).:
```

```html
<div>
    <label class="label focus">Name</label>
    <input type="text">
</div>
```

To add class to the existing ones instead of replacing, use `addLabelClass()` method. In case of multiple classes, pass
them as separate arguments.

```php
echo \Yiisoft\Form\Field\Text::widget()
    ->label('Name')
    ->labelClass('label')
    ->addLabelClass('focus', 'primary');    
```

```html
<div>
    <label class="label focus primary">Name</label>
    <input type="text">
</div>
```

### `hintClass()` / `addHintClass()`

HTML class for hint. In case of multiple classes, pass them as separate arguments.

```php
echo \Yiisoft\Form\Field\Text::widget()
    ->hint('Enter name')
    ->hintClass('hint', 'focus');
```

> > The hint must be visible (use [`hint()`](#hint) or [`hintConfig()](#hintconfig) methods).:

Reviewer: ```suggestion
> The hint must be visible (use [`hint()`](#hint) or [`hintConfig()`](#hintconfig) methods).:
```

```html
<div>
    <div class="hint focus">Enter name</div>
    <input type="text">
</div>
```

To add class to the existing ones instead of replacing, use `addHintClass()` method. In case of multiple classes, pass
them as separate arguments.

```php
echo \Yiisoft\Form\Field\Text::widget()
    ->hint('Name')
    ->hintClass('hint')
    ->addHintClass('focus primary');    
```

```html
<div>
    <input type="text">
    <div class="hint focus primary">Enter name</div>
</div>
```

### `errorClass()` / `addErrorClass()`

HTML class for error. In case of multiple classes, pass them as separate arguments.

```php
echo \Yiisoft\Form\Field\Text::widget()
    ->error('Name is not valid.')
    ->errorClass('error', 'focus');
```

> > The error must be visible (use [`error()`](#error) or [`errorConfig()](#errorconfig) methods).:

Reviewer: ```suggestion
> The error must be visible (use [`error()`](#error) or [`errorConfig()`](#errorconfig) methods).:
```

    ->addInputContainerAttributes(['class' => 'focus', 'data-sort' => 1]);
```

```html
<div>
    <span class="focus" data-type="name" data-sort="1"><input type="text"></span>
</div>
```

### `inputContainerClass()` / `addInputContainerClass()`

HTML class for outer container that wraps the input. Input container tag must be specified for container to be created. 
In case of multiple classes, pass them as separate arguments.

```php
echo \Yiisoft\Form\Field\Text::widget()
    ->inputContainerTag('span')
    ->inputContainerClass('input-container', 'focus');
```

```html
<div>
    <span class="input-container focus"><input type="text"></span>
</div>
```

When not specified, no class is used for input.

To add class to existing ones instead of replacing, use `inputContainerClass()` method. In case of multiple classes, 
pass them as separate arguments.

```php
echo \Yiisoft\Form\Field\Text::widget()
    ->inputContainerTag('span')
    ->inputContainerClass('input-container')
    ->addInputContainerAttributes('focus', 'primary');
```

```html
<div>
    <span class="input-container focus primary"><input type="text"></span>
</div>
```

### `template()`

> A template for a field where tokens (placeholders) are field parts. This template is used when field is created using
> `widget()` method.

Reviewer: ```suggestion
A template for a field where tokens (placeholders) are field parts. This template is used when field is rendered.
```

> Supported tokens:

Reviewer: ```suggestion
Built-in tokens:
```
